### PR TITLE
(PUP-8041) replace FACTER_url with target for puppet resource

### PR DESCRIPTION
**Pull request by tkishel**
Prior to this commit, it was necessary to set FACTER_url when using
puppet resource with a device.

This commit implements a device method in network_device/config.rb,
and uses it in application/device.rb and application/resource.rb.

`puppet resource --target bigip.example.com f5_virtualserver`

This method is also available/useful for device module tasks
that need to target a device.

### CODE REVIEW

**Comment by tkishel** [comment]
Note that 'replace' is misleading: FACTER_url continues to function after this change/commit.

**Comment by shermdog** [comment]
We will have to discuss how we feel about changing the nomenclature of "device / certificate" as that was intentionally chosen to reflect the fact that the device name equals the certificate name.

**Comment by tkishel** [comment]
For the discussion ...

Do we make the same distinction outside the context of devices?
What about updating config_file_device.html as well or instead?

Given:
> Target a specific device/certificate in the device.conf. 
> Doing so will perform a device run against only that device/certificate.

Are there certificates in device.conf?
Do we run against a certificate?

But it's not a hill I want to die on, so feel free to change it or ask me to do the same :)

**Comment by tphoney** [comment]
So basically you have moved the target code out of device, so resource can avail of it. 
Apologies i am still getting my coding goggles on

**Comment by tkishel** [comment]
Yes, your sweet code from PR 5768. For use by `puppet resource` (and potentially, `puppet apply`) and device tasks, instead of having to duplicate the same:

https://github.com/tkishel/f5-puppet/blob/b400af05d7844b4547f3a72e061570487171b98c/tasks/config_save.rb#L40

File: lib/puppet/application/resource.rb [hunk 1]
```diff
@@ -139,6 +148,20 @@ def main
...
+      if options[:target]
+        require 'puppet/util/network_device/config'
+        target_device = Puppet::Util::NetworkDevice::Config.device(options[:target])
+        if target_device.nil?
+          raise _("Invalid target '%{target}'") % { target: options[:target]}
+        end
+        # Facter(:url) is used by the device module loaded via
+        # Puppet::Resource.indirection.search in find_or_save_resources()
+        url = target_device.url.to_s
+        Facter.add(:url) do
```
Comment (joshcooper) [inline lib/puppet/application/resource.rb]: The fact name seems overly generic. If it's only for remote devices, suggest something like `:device_url`, etc

Comment (tkishel) [reply lib/puppet/application/resource.rb]: Yes, the fact name is too generic, but was chosen years ago :( ...

https://puppet.com/blog/managing-f5-big-ip-network-devices-puppet (search for FACTER_url)

... and is implemented in multiple device modules:

https://github.com/f5devcentral/f5-puppet/blob/master/lib/puppet/provider/f5.rb#L16
https://forge.puppet.com/puppetlabs/netscaler#tips-and-tricks

Also, https://tickets.puppetlabs.com/browse/MODULES-1751

Comment (tkishel) [reply lib/puppet/application/resource.rb]: (I tried to identify an alternate mechanism to pass the device object, but was unable navigate indirection.)

**Comment by shermdog** [comment]
@joshcooper I'm going to review this PR as part of a broader discussion with my team.  I've set it as blocked for now as we may want to make some changes.

I will advise shortly.

**Comment by joshcooper** [comment]
Ping @shermdog. Also it looks like there are conflicts with the PR.

**Comment by tkishel** [comment]
Conflict is with (PUP-8244) Remove unused exception object
I'm happy to delete the conflicting '=> detail' if we're going to merge this.

**Comment by shermdog** [comment]
We're going to close this PR in favor of moving both `apply` and `resource` functionality directly into `puppet device`

**Comment by tkishel** [comment]
Le sigh :)

**Pull request closed without merging**